### DISCUSSION TITLE
Lan Mouse Mobile Server

### DISCUSSION THREAD

**Discussion by rohitsangwan01** [feschber/lan-mouse#379] [Announcements]
As discussed [in this comment](https://github.com/feschber/lan-mouse/issues/217#issuecomment-2464659116), here is the LanMouse Mobile Server initial version:

[LanMouse Mobile GitHub Repository](https://github.com/rohitsangwan01/lan-mouse-mobile)

While a few features are still being worked on, it’s overall functioning well. I've tested it with the following setups:

- Android <-> macOS
- iOS <-> macOS

You can download the test APK from the [Releases section](https://github.com/rohitsangwan01/lan-mouse-mobile/releases).

Feel free to check it out!

**Comment by feschber** [comment]
Very cool! Working out of the box :smiley:.

Scrolling is a bit weird (seems like `on_scroll` is activated *after* completing a scroll instead of immediately and for each "step").

I just merged encryption support so you are going to have to add support for DTLS in case you want to adapt this to the latest version.

**Comment by feschber** [comment]
And one more suggestion:

Tap + drag would be a nice feature, as in keep mouse 1 pressed in case it is immediately (within maybe 250ms) followed by a drag action until the drag is finished.

**Comment by feschber** [comment]
I added a link to your repo to the [README](https://github.com/feschber/lan-mouse?tab=readme-ov-file#android--ios)

**Comment by rohitsangwan01** [comment]
@feschber Thankyou for the feedbacks
Will try to apply the suggestions and add encryption support as well, 
i have to check the code on how Encryption is implemented

**Comment by rohitsangwan01** [comment]
Am having issue in implementing keyEvents, i tried implementing like [this](https://github.com/rohitsangwan01/lan-mouse-mobile/blob/bc0e6b4de087835a6d58a5456b85215c73fc6e55/lib/app/modules/server/widgets/keyboard_handler.dart#L19), but after sending these events
MacOs stucked in keyPressed state, and only option i had was to force restart Mac
Seems like KeyRelease event is somehow ignored or maybe some kind of race condition

**Comment by feschber** [comment]
I recently fixed an issue with stuck keys on MacOS concerning modifier keys (CTRL / OPTION / CMD / CAPSLOCK / SHIFT).

Did you check the latest version?

Definitely possible there is still something wrong though

**Comment by feschber** [comment]
For encryption you dont really need to know anything other than the fact that its using DTLS v1.2 and a sha256 fingerprint for certificate verification.

You can obtain a fingerprint through openssl using

```sh
openssl x509 -in lan-mouse.pem -noout -sha256 -fingerprint
```

where `lan-mouse.pem` is the file containing the ssl certificate.

**Comment by rohitsangwan01** [comment]
Thankyou for the info, will implementing in new versions

**Comment by rohitsangwan01** [comment]
@feschber Added Encryption support in [V0.0.2](https://github.com/rohitsangwan01/lan-mouse-mobile/releases/tag/0.0.2)
Udp and Encryption logic is now in Rust and UI part in Flutter

**Comment by feschber** [comment]
That was fast! I'm impressed. And it's working flawlessly as well! I updated the REAMDE accordingly

**Comment by rohitsangwan01** [comment]
Glad to hear it's working smoothly! it's always great when everything comes together seamlessly. Let me know if there’s anything else I can help with!

**Comment by rohitsangwan01** [comment]
@feschber do you think using Flutter for Frontend can be an option for LanMouse ?
There are plugins to easily integrate Flutter and Rust like [flutter_rust_bridge](https://pub.dev/packages/flutter_rust_bridge), which i have also used in LanMouseMobile, and few Popular softwares like RustDesk is also using Rust + Flutter
I think its easy to build on every platform and more freedom in building UI

**Comment by feschber** [comment]
I intentionally designed lan-mouse in a way to make it easy to create new frontends. I don't want to _replace_ the Gtk frontend, but I certainly dont have anything against a flutter frontend.

**Comment by feschber** [comment]
[lan-mouse-ipc](https://github.com/feschber/lan-mouse/tree/main/lan-mouse-ipc) is responsible for the communication, if you want to have a look

**Comment by rohitsangwan01** [comment]
I tried writing frontend in flutter,
It works fine but Issue is to maintain dart and lanMouse's rust communication protocols in sync
In case if anything changes in future

**Comment by mingww64** [comment]
I built an Android TV app that simulates HDMI CEC experience (TV to lan-mouse clients) based on this mobile server. 
[Lan Mouse CEC](https://github.com/mingww64/lan-mouse-tv-cec)
